sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOWriteTest.java

```diff
@@ -341,6 +343,35 @@ private void runTestWrite(String[] elems, String header, String footer) throws E
     runTestWrite(elems, header, footer, 1);
   }
 
+  private static class MatchesFilesystem implements SerializableFunction<Iterable<String>, Void> {
+    private final ResourceId baseFilename;
+
+    MatchesFilesystem(ResourceId baseFilename) {
+      this.baseFilename = baseFilename;
+    }
+
+    @Override
+    public Void apply(Iterable<String> values) {
+      try {
+        String pattern = baseFilename.toString() + "*";
+        Iterable<String> matches =
```

Comment:
* This will look better as a simple loop
* You can put the creation of expected values into runTestWrite, and use regular PAssert.that(collection).contains(those values).

Reply:
runTestWrite doesn't necessarily know the expected values a priori (e.g. if dynamic sharding is used). That's why we match the files written to FS.

Changed to a loop.